src/mast/ast.rs

```diff
@@ -0,0 +1,21 @@
+use crate::{
+    backends::Backend,
+    parser::{Expr, ExprKind},
+};
+
+use super::MastCtx;
+
+impl Expr {
+    /// Convert an expression to another expression, with the same span and a regenerated node id.
+    pub fn to_mast<B: Backend>(&self, ctx: &mut MastCtx<B>, kind: &ExprKind) -> Expr {
+        if !ctx.in_generic_func {
+            return self.clone();
+        }
+
+        Expr {
+            node_id: ctx.next_node_id(),
```

Comment:
New node_id will be generated if the node is within a generic function. When the generic function is instantiated, the new function body is created, thus thus needs to recreate the nodes inside.

Reply:
If it it not within the generic function, then the node id should remain the same as it does `self.clone()`

Reply:
I think when I wrote that comment I was wondering why assign the cloned node a new node id though? IIRC we only use the node id to record the type of a node, and the type should not change unless there was a generic involved with that type (although we might have more generics everywhere in the future so maybe better to leave it like this)

Reply:
The node id is indeed for type tracking.
The main reason I assign new id to the nodes within a generic function, is because they are in a new branch as the result of instantiated function body. 
Although the types behind the node ids may remain the same after function instantiation, I think it is safer to create a new set of node ids to differentiate from the ones in the generic function.